File: Flatten_Binary_Tree_to_Linked_List.cpp
//Time Complexity-O(n)
//Space Complexity-O(1)
//Did the code run on Leetcode? yes

class Solution {
public:
    void flatten(TreeNode* root) {
        TreeNode* cur = root;
        while(cur!=NULL) {
            if(cur->left!=NULL) {
                TreeNode* tmp = cur->left;
                while(tmp->right!=NULL) 
                {
                    tmp = tmp->right;
                }
                    tmp->right = cur->right;
                    cur->right = cur->left;
                    cur->left = nullptr;
            }
            cur = cur->right;
        }
    }
};

File: flatten-binary-tree-to-linked-list.js
// Time Complexity : O(N)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var flatten = function(root) {
    if (!root) return;
    
    let cursor = root;
    while (cursor != null) {
        if (cursor.left) {
            let rightNode = cursor.left;
            while (rightNode.right != null) {
                rightNode = rightNode.right;
            }
            
            rightNode.right = cursor.right;
            cursor.right = cursor.left;
            cursor.left = null;
        }
        cursor = cursor.right;
    }
};
